### PR TITLE
test: add unit test for jobset controller

### PR DESCRIPTION
**Pull request by googs1025**
issue from https://github.com/kubernetes-sigs/jobset/issues/476
It's a bit embarrassing. I accidentally deleted the PR during synchronization and resubmitted a new one. Sorry.
When there is rebase, I don’t know how to do it, and I often close the submission.

### CODE REVIEW

**Comment by danielvegamyhre** [comment]
>When there is rebase, I don’t know how to do it, and I often close the submission.

No worries! I will take a look at this tomorrow.

File: pkg/util/testing/wrappers.go [hunk 1]
```diff
@@ -304,6 +310,24 @@ func (j *JobWrapper) Ready(ready int32) *JobWrapper {
 	return j
 }
 
+// DelTimestamp sets the job deletionTimestamp.
+func (j *JobWrapper) DelTimestamp(deletionTimestamp *metav1.Time) *JobWrapper {
```
Comment (danielvegamyhre) [inline pkg/util/testing/wrappers.go:314]: can we use the full name here please (`DeletionTimestamp`) for consistency across the wrappers?

Same with the resourceVersion below

Comment (googs1025) [reply pkg/util/testing/wrappers.go]: ya, that makes sense

**Comment by googs1025** [comment]
@danielvegamyhre Thank you for your response. I will update the content of this PR after you merge your pr.

File: pkg/controllers/jobset_controller_test.go [hunk 2]
```diff
@@ -1174,6 +1182,1042 @@ func jobWithFailedCondition(name string, failureTime time.Time) *batchv1.Job {
...
+				Delete: func(ctx context.Context, client client.WithWatch,
+					obj client.Object, opts ...client.DeleteOption) error {
+					job, ok := obj.(*batchv1.Job)
+					if !ok {
+						return nil
+					}
+					if tc.errorFlag || job == nil {
+						return errors.New("delete pod error")
+					}
+					job = tc.jobs[0]
```
Comment (danielvegamyhre) [inline pkg/controllers/jobset_controller_test.go:1490]: Why are you overriding the `job` that was deleted with the test case job at index 0 every time?

In the `jobsDeleted` slice, we should just track the actual `job` objects that were in the `Delete()` calls.

File: pkg/controllers/jobset_controller_test.go [hunk 3]
```diff
@@ -1174,6 +1182,1042 @@ func jobWithFailedCondition(name string, failureTime time.Time) *batchv1.Job {
...
+		jobSetName = "test-jobset"
+		ns         = "default"
+	)
+	tests := []struct {
+		name     string
+		js       *jobset.JobSet
+		expected bool
+	}{
+		{
+			name: "test jobSetFinished with not finished",
```
Comment (danielvegamyhre) [inline pkg/controllers/jobset_controller_test.go:1528]:
```suggestion
			name: "jobset not finished",
```

File: pkg/controllers/jobset_controller_test.go [hunk 4]
```diff
@@ -1174,6 +1182,1042 @@ func jobWithFailedCondition(name string, failureTime time.Time) *batchv1.Job {
...
+			js: testutils.MakeJobSet(jobSetName, ns).
+				ReplicatedJob(testutils.MakeReplicatedJob("replicated-job-1").
+					Job(testutils.MakeJobTemplate("test-job", ns).Obj()).
+					Replicas(1).
+					Obj()).Obj(),
+		},
+		{
+			name: "test jobSetFinished with finished",
```
Comment (danielvegamyhre) [inline pkg/controllers/jobset_controller_test.go:1536]:
```suggestion
			name: "jobset finished",
```

File: pkg/controllers/jobset_controller_test.go [hunk 5]
```diff
@@ -1174,6 +1182,1042 @@ func jobWithFailedCondition(name string, failureTime time.Time) *batchv1.Job {
...
+			js: testutils.MakeJobSet(jobSetName, ns).
+				SetConditions([]metav1.Condition{{Type: string(jobset.JobSetCompleted), Status: metav1.ConditionTrue}}).
+				ReplicatedJob(testutils.MakeReplicatedJob("replicated-job-1").
+					Job(testutils.MakeJobTemplate("test-job", ns).Obj()).
+					Replicas(1).
+					Obj()).Obj(),
+			expected: true,
+		},
+	}
+	for _, tc := range tests {
```
Comment (danielvegamyhre) [inline pkg/controllers/jobset_controller_test.go:1546]: Can you add a test case for a failed JobSet? It should be considered finished as well.

File: pkg/controllers/jobset_controller_test.go [hunk 6]
```diff
@@ -1174,6 +1182,1042 @@ func jobWithFailedCondition(name string, failureTime time.Time) *batchv1.Job {
...
+
+func TestResumeJob(t *testing.T) {
+	var (
+		jobSetName = "test-jobset"
+		ns         = "default"
+	)
+	tests := []struct {
+		name                string
+		jobs                []*batchv1.Job
+		expectedJobsUpdated []*batchv1.Job
```
Comment (danielvegamyhre) [inline pkg/controllers/jobset_controller_test.go:1562]:
```suggestion
		expectedResumed []*batchv1.Job
```

File: pkg/controllers/jobset_controller_test.go [hunk 7]
```diff
@@ -1174,6 +1182,1042 @@ func jobWithFailedCondition(name string, failureTime time.Time) *batchv1.Job {
...
+func TestCreateJobs(t *testing.T) {
+	var (
+		jobSetName = "test-jobset"
+		ns         = "default"
+	)
+	tests := []struct {
+		name                string
+		js                  *jobset.JobSet
+		jobs                childJobs
+		expectedErr         error
```
Comment (danielvegamyhre) [inline pkg/controllers/jobset_controller_test.go:1194]: Can we rename `expectedErr` to `wantErr` everywhere in this file?

File: pkg/controllers/jobset_controller_test.go [hunk 8]
```diff
@@ -1174,6 +1182,1042 @@ func jobWithFailedCondition(name string, failureTime time.Time) *batchv1.Job {
...
+		errorFlag           bool
```
Comment (danielvegamyhre) [inline pkg/controllers/jobset_controller_test.go:1195]: Can we rename `errorFlag` to `forceError` everywhere in this file? I think this name makes the parameter's purpose more clear.

File: pkg/controllers/jobset_controller_test.go [hunk 9]
```diff
@@ -1174,6 +1182,1042 @@ func jobWithFailedCondition(name string, failureTime time.Time) *batchv1.Job {
...
+				Create: func(ctx context.Context, client client.WithWatch,
+					obj client.Object, opts ...client.CreateOption) error {
+					job, ok := obj.(*batchv1.Job)
+					if !ok {
+						return nil
+					}
+					if tc.errorFlag || job == nil {
+						return errors.New("create pod error")
+					}
+					job = tc.jobs.active[0]
```
Comment (danielvegamyhre) [inline pkg/controllers/jobset_controller_test.go:1371]: We need to remove this line, should be tracking the `job` passed in the Create() call, not the defined in the test case.

File: pkg/controllers/jobset_controller_test.go [hunk 10]
```diff
@@ -1219,3 +2263,30 @@ func makeJob(args *makeJobArgs) *testutils.JobWrapper {
...
 	return jobWrapper
 }
+
+type fakeClient struct {
+	client client.WithWatch
+	scheme *runtime.Scheme
+	record record.EventRecorder
+}
+
+func makeClient(interceptor interceptor.Funcs, initObjs ...client.Object) *fakeClient {
```
Comment (danielvegamyhre) [inline pkg/controllers/jobset_controller_test.go:2273]:
```suggestion
func makeFakeClient(interceptor interceptor.Funcs, initObjs ...client.Object) *fakeClient {
```

File: pkg/controllers/jobset_controller_test.go [hunk 11]
```diff
@@ -1174,6 +1182,1042 @@ func jobWithFailedCondition(name string, failureTime time.Time) *batchv1.Job {
...
+					Obj()).Obj(),
+			jobs: childJobs{
+				active: []*batchv1.Job{
+					makeJob(&makeJobArgs{
+						jobSetName:        jobSetName,
+						replicatedJobName: "replicated-job-1",
+						jobName:           "test-jobset-replicated-job-1-test-job-0",
+						ns:                ns,
+						replicas:          1,
+						jobIdx:            0,
```
Comment (danielvegamyhre) [inline pkg/controllers/jobset_controller_test.go:1213]: Many of these `makeJob()` calls in the test cases are identical, let's just store 1-2 test jobs as variable above the test cases, then re-use those variables as much as possible in the test cases. This applies to all the tests in this file.

This should drastically reduce the number of lines in this file and improve the readability of the test cases.

**Comment by googs1025** [comment]
Maybe this PR can wait until @danielvegamyhre  these PRs https://github.com/kubernetes-sigs/jobset/pull/496 https://github.com/kubernetes-sigs/jobset/pull/494 are merged before I make new changes.

**Comment by googs1025** [comment]
We can prioritize handling this PR https://github.com/kubernetes-sigs/jobset/pull/490 first.

**Comment by danielvegamyhre** [comment]
> We can prioritize handling this PR #490 first.

Sounds good I will take a look at #490 again

**Comment by danielvegamyhre** [comment]
/hold

until comments addressed

**Comment by danielvegamyhre** [comment]
> @googs1025 I am refactoring createJobs() to make it more testable in #496 since right now this function does a lot things unrelated to creating jobs (enforcing startup policy rules, suspend/resume, etc.) which makes it harder to unit test thoroughly.
> 
> This will make it easier for you to write the tests, and easier for me to review them as well.
> 
> I added some preliminary notes on some of your test in the meantime.

The refactor is complete, feel free to resume this PR

**Comment by googs1025** [comment]
@danielvegamyhre Okay, I plan to spend some time next week working on this PR

**Comment by danielvegamyhre** [comment]
@googs1025 Ok please submit PRs with tests for only 1 function at a time, so that it is easier to review.

**Comment by danielvegamyhre** [comment]
@googs1025 are you still working on this?

**Comment by googs1025** [comment]
> @googs1025 are you still working on this?

Yes, I have been quite busy with work recently. I will start this PR next week.
thanks for reminding

**Comment by danielvegamyhre** [comment]
Just following up on this @googs1025, do you still plan to finish this?

**Comment by googs1025** [comment]
> Just following up on this @googs1025, do you still plan to finish this?
@danielvegamyhre 
Sorry, maybe I can let this issue go, I have other things with higher priority to do

**Pull request closed without merging**